### PR TITLE
3732 upgrade pydantic to v2

### LINKED ISSUE #3732: Upgrade Pydantic from v1 to v2
## Description

Pydantic v1 will be no longer supported as of June 2024. Pydantic v2 has been released and provides features that may allow us to remove certain workarounds we've created.

Upgrade Pydantic from v1 to v2 the Island, Agent, and all plugins.

### PR DESCRIPTION
**Pull request by ilija-lazoroski**
# What does this PR do?

Fixes #3732.

Commits to review:

image

## PR Checklist
* [x] Have you added an explanation of what your changes do and why you'd like to include them?
* [ ] Is the TravisCI build passing?
* [ ] Was the CHANGELOG.md updated to reflect the changes?
* [x] Was the documentation framework updated to reflect the changes?
* [x] Have you checked that you haven't introduced any duplicate code?

## Testing Checklist

* [x] Added relevant unit tests?
* [x] Do all unit tests pass?
* [x] Do all end-to-end tests pass?
* [x] Any other testing performed?
    > Tested by {Running the Monkey locally with relevant config/running Island/...}
* [x] If applicable, add screenshots or log transcripts of the feature working

### CODE REVIEW

File: envs/monkey_zoo/README.md [hunk 1]
```diff
@@ -41,7 +41,8 @@ Example:
 Example:
   `packer build -only googlecompute.mimikatz-15,googlecompute.snmp-20 .`
 
-If you want to keep the machine running to debug the image, add `-debug` flag to the command.
+If you want to disable parallelization and output debugging info, add `-debug` flag to the command.
+If you want to allow editing and retrying of a failed script, use the `-on-error=ask` flag.
```
Comment (VakarisZ) [inline envs/monkey_zoo/README.md:45]: ?

File: envs/monkey_zoo/packer/log4j_logstash_56.pkr.hcl [hunk 2]
```diff
@@ -1,5 +1,5 @@
-source "googlecompute" "log4j-logstash-56-test" {
-  image_name     = "log4j-logstash-56-test"
+source "googlecompute" "log4j-logstash-56" {
+  image_name     = "log4j-logstash-56"
```
Comment (VakarisZ) [inline envs/monkey_zoo/packer/log4j_logstash_56.pkr.hcl:2]: ?

File: monkey/common/agent_configuration/agent_configuration.py [hunk 3]
```diff
@@ -1,13 +1,14 @@
...
-from pydantic import Field, confloat
+from pydantic import Field
+from typing_extensions import Annotated
 
 from .agent_sub_configurations import PolymorphismConfiguration, PropagationConfiguration
 
 
 class AgentConfiguration(MutableInfectionMonkeyBaseModel):
-    keep_tunnel_open_time: confloat(ge=0) = Field(  # type: ignore[valid-type]
+    keep_tunnel_open_time: Annotated[float, Field(ge=0)] = Field(  # type: ignore[valid-type]
```
Comment (VakarisZ) [inline monkey/common/agent_configuration/agent_configuration.py:11]: Do we still need type ignore?

Comment (ilija-lazoroski) [reply monkey/common/agent_configuration/agent_configuration.py]: Nope, removed.

File: monkey/common/agent_events/file_encryption_event.py [hunk 4]
```diff
@@ -30,15 +30,18 @@ class FileEncryptionEvent(AbstractAgentEvent):
         :param error_message: Message if an error occurs during file encryption
     """
 
+    model_config = ConfigDict(arbitrary_types_allowed=True)
```
Comment (VakarisZ) [inline monkey/common/agent_events/file_encryption_event.py:33]: It's for `PurePath`?

File: monkey/common/agent_plugins/agent_plugin.py [hunk 5]
```diff
@@ -28,11 +23,6 @@ class AgentPlugin(InfectionMonkeyBaseModel):
...
 
-    class Config(InfectionMonkeyModelConfig):
-        # b64encode() returns bytes, so we call decode() to transform bytes to str
-        # Pydantic is not able to inherit AgentPluginManifest json encoders
-        # so we update the current encoders
-        json_encoders: Mapping[Type, Callable[[Any], Any]] = {
-            bytes: lambda byte_field: b64encode(byte_field).decode(),
-            **AgentPluginManifest.Config.json_encoders,
-        }
+    @field_serializer("source_archive", when_used="json")
```
Comment (VakarisZ) [inline monkey/common/agent_plugins/agent_plugin.py:26]: Line 22 needs to be addressed. Maybe we can do `JSON[Dict[str, Any]]`?

Comment (ilija-lazoroski) [reply monkey/common/agent_plugins/agent_plugin.py:26]: Nope, I have tried. I don't think that they have solved it yet.

File: monkey/tests/unit_tests/monkey_island/conftest.py [hunk 6]
```diff
@@ -64,7 +64,7 @@ def init_mock_datastore() -> MongoEngineUserDatastore:
     db = MongoEngine()
     db.disconnect(alias="default")
     db_name = insecure_generate_random_string(8)
-    db.connect(db_name, mongo_client_class=mongomock.MongoClient)
+    db.connect(db_name, mongo_client_class=mongomock.MongoClient, uuidRepresentation="pythonLegacy")
```
Comment (cakekoa) [inline monkey/tests/unit_tests/monkey_island/conftest.py:67]: What prompted this change?

File: monkey/monkey_island/Pipfile [hunk 7]
```diff
@@ -18,7 +18,7 @@ ring = ">=0.7.3"
 Flask-RESTful = ">=0.3.8"
 Flask = "==2.2.5"  # v2.3.0 removes JSONEncoder which causes exceptions
 Flask-Limiter = "*"
-Werkzeug = ">=1.0.1"
+Werkzeug = "==2.3.0"
```
[the remark this answers is not in the document]
Comment (ilija-lazoroski) [reply monkey/monkey_island/Pipfile:21]: There was an error with the update of Werkzeug which was only solvable if I pin it to 2.3.0. To be honest, now I don't remember what it was.

File: monkey/infection_monkey/monkey.spec [hunk 8]
```diff
@@ -69,7 +69,7 @@ def get_hidden_imports():
     # to plugins, we can attempt to remove the cryptography dependency from the agent entirely.
     # UPDATE: We can't remove the dependency entirely as doing so causes the Agent to crash.
     #         See https://github.com/guardicore/monkey/issues/3170#issuecomment-1623503645.
-    imports = ['_cffi_backend', '_mssql', 'asyncore', 'logging.config', 'cryptography.hazmat.primitives.padding', 'xml.dom', 'timeit', 'sqlite3']
+    imports = ['_cffi_backend', '_mssql', 'asyncore', 'logging.config', 'cryptography.hazmat.primitives.padding', 'xml.dom',  'xml.etree', 'xml.etree.ElementTree','timeit', 'sqlite3']
```
[the remark this answers is not in the document]
Comment (ilija-lazoroski) [reply monkey/infection_monkey/monkey.spec:72]: I will try

[the remark this answers is not in the document]
Comment (cakekoa) [reply monkey/infection_monkey/monkey.spec:72]: Referencing our past conversations, it affected the powershell exploiter

File: monkey/common/agent_configuration/agent_sub_configurations.py [hunk 9]
```diff
@@ -206,7 +205,7 @@ class PropagationConfiguration(MutableInfectionMonkeyBaseModel):
         :param exploitation: Configuration for exploitation
     """
 
-    maximum_depth: conint(ge=0) = Field(  # type: ignore[valid-type]
+    maximum_depth: Annotated[int, Field(ge=0)] = Field(  # type: ignore[valid-type]
```
Comment (cakekoa) [inline monkey/common/agent_configuration/agent_sub_configurations.py:208]: Should we add this Field to the annotation as well?

**Review by cakekoa — APPROVED** [review]
Nice work!

**Pull request merged**